sw/device/lib/dif/dif_usbdev.c

```diff
@@ -0,0 +1,934 @@
+               .bit_index = USBDEV_IN_SENT_SENT4},
+        [5] = {.config_in_reg_offset = USBDEV_CONFIGIN5_REG_OFFSET,
+               .bit_index = USBDEV_IN_SENT_SENT5},
+        [6] = {.config_in_reg_offset = USBDEV_CONFIGIN6_REG_OFFSET,
+               .bit_index = USBDEV_IN_SENT_SENT6},
+        [7] = {.config_in_reg_offset = USBDEV_CONFIGIN7_REG_OFFSET,
+               .bit_index = USBDEV_IN_SENT_SENT7},
+        [8] = {.config_in_reg_offset = USBDEV_CONFIGIN8_REG_OFFSET,
+               .bit_index = USBDEV_IN_SENT_SENT8},
+        [9] = {.config_in_reg_offset = USBDEV_CONFIGIN9_REG_OFFSET,
+               .bit_index = USBDEV_IN_SENT_SENT9},
+        [10] = {.config_in_reg_offset = USBDEV_CONFIGIN10_REG_OFFSET,
+                .bit_index = USBDEV_IN_SENT_SENT10},
+        [11] = {.config_in_reg_offset = USBDEV_CONFIGIN11_REG_OFFSET,
+                .bit_index = USBDEV_IN_SENT_SENT11},
+};
+
+static bool is_valid_endpoint(uint32_t endpoint) {
+  return endpoint < USBDEV_NUM_ENDPOINTS;
+}
+
+/**
+ * Mapping from `dif_usbdev_irq_t` to bit indices in `USBDEV_INTR_STATE`
+ * register.
+ */
+static const uint8_t kIrqEnumToBitIndex[] = {
+        [kDifUsbdevIrqPktReceived] = USBDEV_INTR_STATE_PKT_RECEIVED,
+        [kDifUsbdevIrqPktSent] = USBDEV_INTR_STATE_PKT_SENT,
+        [kDifUsbdevIrqDisconnected] = USBDEV_INTR_STATE_DISCONNECTED,
+        [kDifUsbdevIrqHostLost] = USBDEV_INTR_STATE_HOST_LOST,
+        [kDifUsbdevIrqLinkReset] = USBDEV_INTR_STATE_LINK_RESET,
+        [kDifUsbdevIrqLinkSuspend] = USBDEV_INTR_STATE_LINK_SUSPEND,
+        [kDifUsbdevIrqLinkResume] = USBDEV_INTR_STATE_LINK_RESUME,
+        [kDifUsbdevIrqAvEmpty] = USBDEV_INTR_STATE_AV_EMPTY,
+        [kDifUsbdevIrqRxFull] = USBDEV_INTR_STATE_RX_FULL,
+        [kDifUsbdevIrqAvOverflow] = USBDEV_INTR_STATE_AV_OVERFLOW,
+        [kDifUsbdevIrqLinkInError] = USBDEV_INTR_STATE_LINK_IN_ERR,
+        [kDifUsbdevIrqRxCrcError] = USBDEV_INTR_STATE_RX_CRC_ERR,
+        [kDifUsbdevIrqRxPidError] = USBDEV_INTR_STATE_RX_PID_ERR,
+        [kDifUsbdevIrqRxBitstuffError] = USBDEV_INTR_STATE_RX_BITSTUFF_ERR,
+        [kDifUsbdevIrqFrame] = USBDEV_INTR_STATE_FRAME,
+        [kDifUsbdevIrqConnected] = USBDEV_INTR_STATE_CONNECTED,
+};
+
+static bool is_valid_irq(dif_usbdev_irq_t irq) {
+  return irq >= kDifUsbdevIrqFirst && irq <= kDifUsbdevIrqLast;
+}
+
+/**
+ * Static functions for the free buffer pool.
+ */
+
+static bool buffer_pool_is_full(dif_usbdev_buffer_pool_t *pool) {
+  return pool->top == (USBDEV_NUM_BUFFERS - 1);
+}
+
+static bool buffer_pool_is_empty(dif_usbdev_buffer_pool_t *pool) {
+  return pool->top == -1;
+}
+
+static bool buffer_pool_is_valid_entry(uint8_t entry) {
+  return entry < USBDEV_NUM_BUFFERS;
+}
+
+static bool buffer_pool_add(dif_usbdev_buffer_pool_t *pool, uint8_t entry) {
+  if (buffer_pool_is_full(pool) || !buffer_pool_is_valid_entry(entry)) {
+    return false;
+  }
+
+  pool->entries[++pool->top] = entry;
```

Comment:
Maybe split out the `++` into another statement? Most people (myself included) might not remember the precedence of `++` vs `->`.

(In general, I would personally avoid using `++foo` as an expression.)